In [3]:
import numpy as np

# Arrays and Lists

In [5]:
L = [1, 2, 3]

In [6]:
A = np.array([1, 2, 3])

In [7]:
for e in L:
    print(e)


1
2
3


In [8]:
for e in A:
    print(e)

1
2
3


In [9]:
L.append(4)

In [10]:
L

[1, 2, 3, 4]

In [11]:
L + [5]

[1, 2, 3, 4, 5]

In [12]:
A

array([1, 2, 3])

In [13]:
A + np.array([4])

array([5, 6, 7])

In [14]:
A + np.array([4, 5, 6])

array([5, 7, 9])

In [15]:
A + np.array([4, 5])

ValueError: operands could not be broadcast together with shapes (3,) (2,) 

In [16]:
2 * A

array([2, 4, 6])

In [17]:
2 * L

[1, 2, 3, 4, 1, 2, 3, 4]

In [18]:
L + L

[1, 2, 3, 4, 1, 2, 3, 4]

In [19]:
L2 = []
for e in L:
    L2.append(e + 3)
L2

[4, 5, 6, 7]

In [20]:
L2 = [e + 3 for e in L]
L2

[4, 5, 6, 7]

In [21]:
L**2

TypeError: unsupported operand type(s) for ** or pow(): 'list' and 'int'

In [22]:
L2 = [e**2 for e in L]
L2

[1, 4, 9, 16]

In [23]:
A**2

array([1, 4, 9])

In [24]:
np.sqrt(A)

array([1.        , 1.41421356, 1.73205081])

In [25]:
np.log(A)

array([0.        , 0.69314718, 1.09861229])

In [26]:
np.exp(A)

array([ 2.71828183,  7.3890561 , 20.08553692])

In [27]:
np.tanh(A)

array([0.76159416, 0.96402758, 0.99505475])

# The Dot Product

$$ a \cdot b = a^T b = \sum_{d=1}^D a_d b_d $$

In [28]:
a = np.array([1, 2])
b = np.array([3, 4])

In [29]:
dot = 0
for e, f in zip(a, b):
    dot += e * f
dot

11

In [30]:
dot = 0
for i in range(len(a)):
    dot += a[i] * b[i]
dot

11

In [31]:
a * b

array([3, 8])

In [32]:
np.sum(a * b)

11

In [33]:
(a * b).sum()

11

In [34]:
np.dot(a, b)

11

In [35]:
a.dot(b)

11

In [36]:
a @ b

11

$$ a^T b = \left\lVert a \right\rVert \left\lVert b \right\rVert \cos \theta_{ab}$$

$$ \cos \theta_{ab} = \frac{a^T b}{\left\lVert a \right\rVert \left\lVert b \right\rVert} $$

$$ \left\lVert a \right\rVert = \sqrt{\sum_{d=1}^D a_d^2 } $$

In [39]:
amag = np.sqrt((a * a).sum())
amag

2.23606797749979

In [40]:
np.linalg.norm(a)

2.23606797749979

In [48]:
cosangle = a.dot(b) / (np.linalg.norm(a) * np.linalg.norm(b))

In [49]:
cosangle

0.9838699100999074

In [50]:
angle = np.arccos(cosangle)
angle

0.17985349979247847

# Speed Test

In [82]:
## speed comparison ##
import numpy as np
from datetime import datetime

# note: you can also use %timeit

a = np.random.randn(100)
b = np.random.randn(100)
T = 100000

def slow_dot_product(a, b):
  result = 0
  for e, f in zip(a, b):
    result += e*f
  return result

t0 = datetime.now()
for t in range(T):
  slow_dot_product(a, b)
dt1 = datetime.now() - t0

t0 = datetime.now()
for t in range(T):
  a.dot(b)
dt2 = datetime.now() - t0

print("dt1 / dt2:", dt1.total_seconds() / dt2.total_seconds())

dt1 / dt2: 35.35720593429883


# Matrices

In [52]:
L = [[1, 2], [3, 4]]
L

[[1, 2], [3, 4]]

In [54]:
L[0]
L[0][1]

2

In [55]:
A = np.array([[1, 2], [3, 4]])
A

array([[1, 2],
       [3, 4]])

In [56]:
A[0][1]

2

In [57]:
A[0, 1]

2

Select Column

In [58]:
A[:, 0]

array([1, 3])

In [59]:
A.T

array([[1, 3],
       [2, 4]])

In [60]:
np.exp(A)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [61]:
np.exp(L)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [62]:
B = np.array([[1, 2, 3], [4, 5, 6]])
B

array([[1, 2, 3],
       [4, 5, 6]])

In [63]:
A.dot(B)

array([[ 9, 12, 15],
       [19, 26, 33]])

In [64]:
A.dot(B.T)

ValueError: shapes (2,2) and (3,2) not aligned: 2 (dim 1) != 3 (dim 0)

In [65]:
np.linalg.det(A)

-2.0000000000000004

In [66]:
np.linalg.inv(A)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [67]:
np.linalg.inv(A).dot(A)

array([[1.00000000e+00, 0.00000000e+00],
       [1.11022302e-16, 1.00000000e+00]])

In [68]:
np.trace(A)

5

In [69]:
np.diag(A)

array([1, 4])

In [70]:
np.diag([1,4])

array([[1, 0],
       [0, 4]])

In [71]:
np.linalg.eig(A)

(array([-0.37228132,  5.37228132]),
 array([[-0.82456484, -0.41597356],
        [ 0.56576746, -0.90937671]]))

In [72]:
Lam, V = np.linalg.eig(A)

In [73]:
V[:, 0] * Lam[0] == A @ V[:, 0]

array([ True, False])

In [75]:
V[:, 0] * Lam[0], A @ V[:, 0]

(array([ 0.30697009, -0.21062466]), array([ 0.30697009, -0.21062466]))

In [77]:
np.allclose(V[:, 0] * Lam[0], A @ V[:, 0])

True

In [78]:
np.allclose(V @ np.diag(Lam), A @ V)

True

# Solving Linear Systems

The admission fee at a small fair is $1.50 for children and $4.00 for adults.
On a certain day, 2200 people enter the fair, and $5050 is collected.  How many
children and how many adults attended?

2 equations, 2 unknown

$$ x_1 + x_2 = 2200 $$
$$ 1.5x_1 + 4x_2 = 5050 $$

# Linear System in Matrix Form

$$ x = \begin{pmatrix} x_1\\ x_2 \end{pmatrix}, A = \begin{pmatrix} 1 & 1\\ 1.5 & 4 \end{pmatrix}, b = \begin{pmatrix} 2200\\ 5050 \end{pmatrix} $$

$$ Ax = b \Leftrightarrow x = A^{-1}b $$

# Don't do that literally!

- The "inverse" is slower and less accurate
- There are better algorithms to **solve** linear systems

In [85]:
A = np.array([[1, 1], [1.5, 4]])
A

array([[1. , 1. ],
       [1.5, 4. ]])

In [86]:
b = np.array([2200, 5050])
b

array([2200, 5050])

In [88]:
x = np.linalg.solve(A, b) # yes
x

array([1500.,  700.])

In [89]:
x = np.linalg.inv(A).dot(b) # no
x

array([1500.,  700.])

# Generating Data

In [90]:
np.zeros((2, 3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [91]:
np.ones((2, 3))

array([[1., 1., 1.],
       [1., 1., 1.]])

In [92]:
10 * np.ones((2, 3))

array([[10., 10., 10.],
       [10., 10., 10.]])

In [93]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [94]:
np.random.random()

0.14588949373028337

In [95]:
np.random.random((2, 3))

array([[0.3770103 , 0.56907103, 0.11618271],
       [0.29989348, 0.83146184, 0.80397198]])

In [97]:
np.random.randn(2, 3) # from standard normal distribution, mean = 0, variance = 1

array([[ 0.03574672,  0.23661411, -1.40725625],
       [-0.13808613, -0.85832732,  1.04725893]])

In [98]:
R = np.random.randn(10000)

In [99]:
R.mean()

-0.010672663523385712

In [100]:
np.mean(R)

-0.010672663523385712

In [101]:
R.var()

0.9932235046455328

In [102]:
R.std()

0.996605992679922

In [103]:
R = np.random.randn(10000, 3)

In [104]:
R.mean(axis=0) # mean of column

array([ 0.00580234, -0.01311956,  0.00041281])

In [105]:
R.mean(axis=1).shape

(10000,)

In [106]:
np.cov(R).shape # covariance

(10000, 10000)

In [107]:
np.cov(R.T)

array([[ 0.99428372, -0.00547817, -0.00201575],
       [-0.00547817,  1.01415627,  0.01243761],
       [-0.00201575,  0.01243761,  0.9947012 ]])

In [108]:
np.cov(R, rowvar=False)

array([[ 0.99428372, -0.00547817, -0.00201575],
       [-0.00547817,  1.01415627,  0.01243761],
       [-0.00201575,  0.01243761,  0.9947012 ]])

In [109]:
np.random.randint(0, 10, size=(3, 3))

array([[5, 6, 0],
       [7, 9, 3],
       [0, 2, 1]])

In [110]:
np.random.choice(10, size=(3, 3))

array([[1, 9, 8],
       [9, 6, 1],
       [7, 5, 6]])